doc/photoshare.1.md
photoshare (1)
==================

NAME
----

photoshare - Manage a list of which photos (or other files) you've shared,
using metapho's tagging system and the special tag 'share'.

SYNOPSIS
--------

photoshare  
Search in the current directory (or directories specified with -d)
for images that have not been shared.

photoshare img2.jpg img.jpg ...  
Check whether the specified images have been shared, and when

photoshare add img2.jpg img.jpg ...  
Add the specified images to the share list with today's date


OPTIONAL FLAGS
--------------

-----
|   |   |
|---|---|
| -h, --help  | show this help message and exit |
| -d DIRLIST  | Directories to search, comma-separated |



AUTHOR
------

Akkana Peck.

COPYRIGHT
---------

Copyright (C) 2024 Akkana Peck.
Part of Metapho, which is free software,
licensed under the GNU Public License version 2 (or later).

SEE ALSO
--------

The most recent version of `metapho`, and more information about it, is at:
https://github.com/akkana/metapho


CHANGELOG
2024-08-16  Akkana Peck  <akkana@shallowsky.com>

	Add new photoshare script, and beef up the documentation
	on metapho's (now three) scripts.

2023-12-27  Akkana Peck  <akkana@shallowsky.com>

	1.1b3: Show filename and number/total in titlebar.
	Also move g_image_list out of the Image class (now it's
	directly in the metapho module) and provide some functions
	to show total plus total shown and hidden.

2022-04-04  Akkana Peck  <akkana@shallowsky.com>

	fotogr: add a -t, taglines, option
	to print all matching taglines found in Tags files,
	which can guide the next search.

2021-10-09  Akkana Peck  <akkana@shallowsky.com>

	fotogr: allow -d2020 as well as -d 2020,
	and give an error for unrecognized tags.

2021-06-12  Akkana Peck  <akkana@shallowsky.com>

	1.1b2: Add fotogr tests.

2021-06-12  Akkana Peck  <akkana@shallowsky.com>

	1.1b1: Add fotogr (moved from scripts repository).

2021-04-29  Akkana Peck  <akkana@shallowsky.com>

	Colorize searches -- that's something that had been lost
	in the move to GTK3.
	Some bug fixes involving tags added mistakenly.

2020-06-26  Akkana Peck  <akkana@shallowsky.com>

	Make the notags script smarter: add a -h for help,
	improve handling of skipped extensions and directories
	and make that user-configurable, and show paths as relative.

2020-04-13  Akkana Peck  <akkana@shallowsky.com>
	Bump version to 1.0 and release on PyPI.

2019-12-22  Akkana Peck  <akkana@shallowsky.com>

	Update main and ImageViewer classes to use gi gtk3 explicitly,
	and to use cairo methods for checking window size.
	Improvements to the image viewer classes.
	 Add resize handling to the viewer to rescale the image
	when window size changes.
	Allow moving forward/back in image lists in the viewer window.
	Handle fullsize/unfullsize events, including resizing the containing
	window if it allows it; implement next/previous so you can have
	multiple images stored in the same zoomview.

2019-12-20  Akkana Peck  <akkana@shallowsky.com>

	1.0b2:
	Port to Python3.

2017-07-30  Akkana Peck  <akkana@shallowsky.com>

	Fix some bugs in ImageViewer involving rotation.

2017-03-03  Akkana Peck  <akkana@shallowsky.com>

	Add a script entry for mpiv image viewer.
	Fix lots of problems related to using multiple categories at once.
	Sort images alphabetically (within a tag) in the Tags file.
	Replace separate notags script with code in metapho.py.

2017-02-20  Akkana Peck  <akkana@shallowsky.com>

	Split ImageViewer off into a new file with __main__.
	Add a new class ImageViewerWindow to allow testing ImageViewer
	and using it as a standalone image viewer, to
	allow viewing zoomed images from within metapho,
	on demand, without needing to copy the current filename and
	start up an external image viewer.


2016-12-16  Akkana Peck  <akkana@shallowsky.com>

	0.6:
	Add a new class CategoryViewer and improve the category UI.

2016-12-15  Akkana Peck  <akkana@shallowsky.com>

	0.5.1:
	Split up gtkpho.py and move gtkpho into a separate directory.

2016-12-03  Akkana Peck  <akkana@shallowsky.com>

	0.5:
	First PyPI version.

2014-06-16  Akkana Peck  <akkana@shallowsky.com>

	Add category support.

2013-02-09  Akkana Peck  <akkana@shallowsky.com>

	First usable version.


metapho/scripts/photoshare.py
#!/usr/bin/env python3

# Keeping track of photo sharing, e.g. on mastodon.

# List of images already shared are stored in SHAREFILE
# as YYYY-MM-DD<TAB>img1.jpg img2.jpg ...

# Search directories containing photos are expected to have a Tags
# file (such as those produced by metapho) in which files tagged
# for sharing have the tag "share".

# Copyright 2024 by Akkana Peck: share and enjoy under the GPLv2 or later.


from datetime import date
import os, sys

from . import fotogr


SHAREFILE = os.path.expanduser('~/Docs/Lists/sharephotos')


def read_in_sharefile():
    sharedfiles = []
    try:
        with open(SHAREFILE) as fp:
            for line in fp:
                datestr, files = line.strip().split('\t')
                filelist = [ f.strip() for f in files.split() ]
                sharedfiles.append((datestr, filelist))
    except FileNotFoundError:
        print("No sharefile found at", SHAREFILE)

    return sharedfiles


def add_to_sharefile(imglist):
    with open(SHAREFILE, 'a') as fp:
        print('%s\t%s' % (date.today().strftime("%Y-%m-%d"),
                          ' '.join(imglist)), file=fp)
        print("Added to the sharefile", SHAREFILE)


def search_in_sharefile(imgname, sharedfiles):
    """The sharefile can have directory fragments, like lassen/img.jpg
       so search for anything that includes namefrag.
       Return the most recent date it was shared, or None if not found.
    """
    namefrag = os.path.basename(imgname)
    most_recent_date = None
    for datestr, filelist in sharedfiles:
        for f in filelist:
            if namefrag in f:
                most_recent_date = datestr
                # print(imgname, "was shared on", datestr,
                #       "because", namefrag, "is in", f)

    return most_recent_date


def find_unshared_images(dirlist, images_to_check):
    shared = []
    unshared = []

    # Read in the sharefile
    sharedfiles = read_in_sharefile()

    if not images_to_check:
        images_to_check = fotogr.search_for_keywords(dirlist, ['share'],
                                                     [], [],
                                                     True, False)

    for img in images_to_check:
        # print(img)
        datestr = search_in_sharefile(img, sharedfiles)
        if datestr:
            shared.append((datestr, img))
        else:
            unshared.append(img)

    return shared, unshared


def main():
    import argparse
    parser = argparse.ArgumentParser(description="",
                                     formatter_class=argparse.RawTextHelpFormatter,
                                     usage="""%(prog)s: Manage shareable photos.
    Search in . (or in other directories given by -d) for photos tagged
    with 'share' that haven't already been shared.

%(prog)s img2.jpg img.jpg ...
    Check whether the specified images have been shared, and when

%(prog)s add img2.jpg img.jpg ...
    Add the specified images to the share list with today's date
""")
    parser.add_argument('-d', action="store", dest="dirlist",
                        help='Directories to search, comma-separated')
    args, rest = parser.parse_known_args(sys.argv)

    if not args.dirlist:
        args.dirlist = '.'

    if len(rest) == 1 or rest[1] != 'add':
        # The most common case, do a search
        shared, unshared = find_unshared_images(args.dirlist, rest[1:])
        for datestr, img in shared:
            print(img, "was shared on", datestr)
        print("unshared images:", ' '.join(unshared))
        sys.exit(0)

    # Add files to the sharefile
    if len(rest) <= 2:
        print("Add what files?")
        sys.exit(1)
    add_to_sharefile(rest[2:])
    sys.exit(0)


if __name__ == '__main__':
    main()
